<?xml version="1.0"?>
<!-- Bringup file to spawn both the world and the 2 drones -->
<launch>
  
    <!-- Flags to select GUI, frame of reference, vehicle and world-->
    <arg name="gui" default="true"/>
    <arg name="world_frame" default="world"/>
    <arg name="world" default="$(find drone_worlds)/worlds/empty.world"/>
    <arg name="use_sim_time" default="true"/>

    <!-- Shuttle's name and identifier -->
    <arg name="nameShuttle" default="typhoon_h480"/>
    <arg name="IDShuttle" default="0"/>
    
    <!-- PX4 Configurations -->
    <arg name="px4_configShuttle" default="rcS"/>
    <arg name="px4_name_configShuttle" default="typhoon_h480" />

    <!-- Vehicle's initial pose in ENU-->
    <arg name="xShuttle"     default="0.0"/>
    <arg name="yShuttle"     default="0.0"/>
    <arg name="zShuttle"     default="0.0"/>

    <arg name="rollShuttle"  default="0.0"/>
    <arg name="pitchShuttle" default="0.0"/>
    <arg name="yawShuttle"   default="0.0"/>
    

    
    <!-- NOTE: Name defines the model to use, make sure that the corresponding model is available -->
    <!-- FEITO AINDA COM O TYPHOON PARA TESTES COOPERATIVOS; MAIS TARDE MUDAR PARA O PLANE -->

    <!-- Target's name and identifier -->
    <arg name="nameTarget" default="plane"/>
    <arg name="IDTarget" default="1"/>
    
    <!-- PX4 Configurations -->
    <arg name="px4_configTarget" default="rcS"/>
    <arg name="px4_name_configTarget" default="plane" />

    <!-- Vehicle's initial pose in ENU-->
    <arg name="xTarget"     default="0.0"/>
    <arg name="yTarget"     default="5.0"/>
    <arg name="zTarget"     default="0.0"/>

    <arg name="rollTarget"  default="0.0"/>
    <arg name="pitchTarget" default="0.0"/>
    <arg name="yawTarget"   default="0.0"/>
    

    <!-- Relative initial distance from shuttle to target pose in ENU-->

    <param name="initialRelativeDistanceX" type="double" value="$(eval arg('xTarget') - arg('xShuttle'))"/>
    <param name="initialRelativeDistanceY" type="double" value="$(eval arg('yTarget') - arg('yShuttle'))"/>
    <param name="initialRelativeDistanceZ" type="double" value="$(eval arg('zTarget') - arg('zShuttle'))"/>

    <!-- Gazebo sim -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="gui" value="$(arg gui)"/>
        <arg name="world_name" value="$(arg world)"/>
        <arg name="debug" value="false"/>
        <arg name="verbose" value="false"/>
        <arg name="paused" value="false"/>
        <arg name="respawn_gazebo" value="false"/>
    </include>

    <!-- Launch the drone vehicle -->
    <include file="$(find cooperative_planning)/launch/vehicle_bringup.launch">
        <!-- Shuttle -->
        <!-- Vehicle name and Identifier -->
        <arg name="nameShuttle" value="$(arg nameShuttle)"/>
        <arg name="IDShuttle" value="$(arg IDShuttle)" />

        <!-- Initial pose of the vehicle-->
        <arg name="xShuttle" value="$(arg xShuttle)"/>
        <arg name="yShuttle" value="$(arg yShuttle)"/>
        <arg name="zShuttle" value="$(arg zShuttle)"/>
        <arg name="rollShuttle"  value="$(arg rollShuttle)"/>
        <arg name="pitchShuttle" value="$(arg pitchShuttle)"/>
        <arg name="yawShuttle"   value="$(arg yawShuttle)"/>

        <!-- Configurations for the PX4 -->
        <arg name="px4_configShuttle" value="$(arg px4_configShuttle)" />
	    <arg name="px4_name_configShuttle" value="$(arg px4_name_configShuttle)" />



        <!-- Target -->
         <!-- Vehicle name and Identifier -->
        <arg name="nameTarget" value="$(arg nameTarget)"/>
        <arg name="IDTarget" value="$(arg IDTarget)" />

        <!-- Initial pose of the vehicle-->
        <arg name="xTarget" value="$(arg xTarget)"/>
        <arg name="yTarget" value="$(arg yTarget)"/>
        <arg name="zTarget" value="$(arg zTarget)"/>
        <arg name="rollTarget"  value="$(arg rollTarget)"/>
        <arg name="pitchTarget" value="$(arg pitchTarget)"/>
        <arg name="yawTarget"   value="$(arg yawTarget)"/>

        <!-- Configurations for the PX4 -->
        <arg name="px4_configTarget" value="$(arg px4_configTarget)" />
	    <arg name="px4_name_configTarget" value="$(arg px4_name_configTarget)" />
    </include>






    
    <arg name="mission" default="simulation"/>
   

    <!-- Define the namespace for this drone being launched -->
    <arg name="namespaceShuttle" value="$(arg nameShuttle)$(arg IDShuttle)"/>
    <arg name="namespaceTarget" value="$(arg nameTarget)$(arg IDTarget)"/>

    <!-- Load global vehicle parameters -->
    <rosparam command="load" file="$(find drone_bringup)/config/$(arg mission)/$(arg nameShuttle)/config.yaml"/>
    <param name="IDShuttle" type="int" value="$(arg IDShuttle)"/>
    <param name="namespaceShuttle" type="string" value="$(arg namespaceShuttle)"/>
    <param name="IDTarget" type="int" value="$(arg IDTarget)"/>
    <param name="namespaceTarget" type="string" value="$(arg namespaceTarget)"/>

    <node pkg="cooperative_planning" type="offboard_shuttle_follower_controller" name="controller"/> 
    


    
</launch>
